### PR TITLE
AMD GPUPerfAPI integration for Direct3D12, OpenGL, and Vulkan.

### PR DESCRIPTION
**Pull request by herb-amd**
AMD GPUPerfAPI 3.0 Preview integration.
Direct3D Shader bug fix due to missing members.
Windows project fix to place qwindows.dll in the expected output directory, otherwise it's not found.
Fix Direct3D11 counters to actually serialize between commands.

### CODE REVIEW

File: qrenderdoc/qrenderdoc_local.vcxproj [hunk 1]
```diff
@@ -1646,7 +1646,7 @@ IF %ERRORLEVEL% NEQ 0 (echo ====================================================
       <Visible>false</Visible>
     </Content>
     <Content Include="3rdparty\qt\$(Platform)\plugins\platforms\qwindows.dll">
-      <Link>qtplugins\platforms\qwindows.dll</Link>
+      <Link>platforms\qwindows.dll</Link>
```
Comment (baldurk) [inline qrenderdoc/qrenderdoc_local.vcxproj:1649]: Can you explain this change? I'm not sure why it should be necessary. The `qt.conf` in qrenderdoc points the plugins folder at `qtplugins` and the existing setup has certainly worked for everyone who's tried it in local or nightly builds up until now.

File: renderdoc/data/hlsl/debugdisplay.hlsl [hunk 2]
```diff
@@ -199,6 +199,8 @@ struct wireframeV2F
 {
 	float4 pos : SV_Position;
 	float3 norm : Normal;
+	float3 color : COLOR;
+	float2 tex : TEXCOORD0;
```
Comment (baldurk) [inline renderdoc/data/hlsl/debugdisplay.hlsl:203]: What's the reason for this change?

File: renderdoc/driver/d3d11/d3d11_replay.cpp [hunk 3]
```diff
@@ -149,7 +149,10 @@ void D3D11Replay::CreateResources()
   RenderDoc::Inst().SetProgress(LoadProgress::DebugManagerInit, 0.9f);
 
   AMDCounters *counters = new AMDCounters();
-  if(counters->Init((void *)m_pDevice))
+
+  auto d3dDevice = m_pDevice->GetReal();
```
Comment (baldurk) [inline renderdoc/driver/d3d11/d3d11_replay.cpp]: Please avoid using `auto` except in cases where the type is awkward like iterators or lambdas, prefer to expand out the actual type in cases like this.

File: renderdoc/driver/d3d12/d3d12_counters.cpp [hunk 4]
```diff
@@ -167,6 +190,125 @@ CounterDescription D3D12Replay::DescribeCounter(GPUCounter counterID)
...
+    }
+  }
+
+  void PostRedraw(uint32_t eid, ID3D12GraphicsCommandList *cmd) override {}
+  // we don't need to distinguish, call the Draw functions
+  void PreDispatch(uint32_t eid, ID3D12GraphicsCommandList *cmd) override { PreDraw(eid, cmd); }
+  bool PostDispatch(uint32_t eid, ID3D12GraphicsCommandList *cmd) override { return PostDraw(eid, cmd); }
+  void PostRedispatch(uint32_t eid, ID3D12GraphicsCommandList *cmd) override { PostRedraw(eid, cmd); }
+
+  void AliasEvent(uint32_t primary, uint32_t alias) override {}
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_counters.cpp]: Without an implementation for `AliasEvent`, if the same command buffer is submitted to a queue multiple times then subsequent submits won't have any data. For regular counters the solution is just to store the list of aliases and duplicate the values - obviously the values may not come out precisely identical but at least it's better than nothing.

The ideal would be to record it two ways recording two sets of counter data, once for each submit, but there's currently no infrastructure to do so.

File: renderdoc/driver/d3d12/d3d12_counters.cpp [hunk 5]
```diff
@@ -167,6 +190,125 @@ CounterDescription D3D12Replay::DescribeCounter(GPUCounter counterID)
...
+
+  virtual ~D3D12AMDDrawCallback() { m_pDevice->GetQueue()->GetCommandData()->m_DrawcallCallback = NULL; }
+
+  void PreDraw(uint32_t eid, ID3D12GraphicsCommandList *cmd) override
+  {
+    m_pEventIds->push_back(eid);
+
+    WrappedID3D12GraphicsCommandList *pWrappedCmdList = (WrappedID3D12GraphicsCommandList *)cmd;
+
+    if (m_begunCommandLists.find(pWrappedCmdList->GetReal()) == m_begunCommandLists.end())
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_counters.cpp]: Would it be helpful here instead to have another new callback similar to `PreCloseCommandList` but instead `PostResetCommandList` to perform this opening preamble? That shouldn't be hard to add, either I can do it or you.

File: renderdoc/driver/d3d12/d3d12_counters.cpp [hunk 6]
```diff
@@ -241,6 +384,36 @@ vector<CounterResult> D3D12Replay::FetchCounters(const vector<GPUCounter> &count
...
+  std::copy_if(counters.begin(), counters.end(), std::back_inserter(d3dCounters),
+               [](const GPUCounter &c) { return c < GPUCounter::FirstAMD; });
+
+  if(m_pAMDCounters)
+  {
+    // Filter out the AMD counters
+    vector<GPUCounter> amdCounters;
+    std::copy_if(
+        counters.begin(), counters.end(), std::back_inserter(amdCounters),
+        [](const GPUCounter &c) { return c >= GPUCounter::FirstAMD && c < GPUCounter::FirstIntel; });
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_counters.cpp:451]: FYI - I added a `IsAMDCounter` that does this more clearly just in case the enum order changes or something.

File: renderdoc/driver/d3d12/d3d12_replay.cpp [hunk 7]
```diff
@@ -94,6 +95,27 @@ void D3D12Replay::CreateResources()
     m_PixelPick.Init(m_pDevice, m_DebugManager);
     m_Histogram.Init(m_pDevice, m_DebugManager);
   }
+
+  if (RenderDoc::Inst().IsReplayApp())
+  {
+      AMDCounters *counters = new AMDCounters();
+
+      auto d3dDevice = m_pDevice->GetReal();
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_replay.cpp]: `auto` here again :smile:.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 8]
```diff
@@ -22,70 +22,202 @@
...
+  }
+
+#if ENABLED(RDOC_DEVEL)
+#if ENABLED(RDOC_WIN32)
+  ::OutputDebugStringA(message.c_str());
+  ::OutputDebugStringA("\n");
+#else
+  std::cerr << message.c_str() << "\n";
+#endif
+#endif
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp:84]: Please don't add custom logging code here, this is handled by the `RDCWARN`/`RDCLOG` calls.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 9]
```diff
@@ -22,70 +22,202 @@
...
 }
 
-bool AMDCounters::Init(void *pContext)
+std::string AMDCounters::FormatErrMessage(const char* operation, uint32_t status)
 {
-  const char *dllName = "GPUPerfAPIDX11-x64.dll";
+  std::ostringstream o;
+  o << operation << ". " << m_pGPUPerfAPI->GPA_GetStatusAsStr((GPA_Status)status);
+  std::string err = o.str();
+  return err;
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp]: This should be replaced by using the existing formatting function `StringFormat::Fmt` that's used elsewhere in the code please. Ideally we could solve this with adding `ToStr` support for GPA_Status, but the need for `m_pGPUPerfAPI` probably shoots that idea down.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 10]
```diff
@@ -22,70 +22,202 @@
...
+#endif
+
 
+
+  GPA_GetFuncTablePtrType getFuncTable = NULL;
+
+#if ENABLED(RDOC_WIN32)
   // first try in the plugin location it will be in distributed builds
-  HMODULE module = LoadLibraryA(LocatePluginFile("amd/counters", dllName).c_str());
+  HMODULE module = LoadLibraryA(LocatePluginFile("amd/counters", dllName.c_str()).c_str());
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp:127]: This part (after constructing dllName) should use the existing OS-agnostic abstraction layer, `Process::LoadModule()` and `Process::GetFunctionAddress`, rather than having OS-specific functions here.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 11]
```diff
@@ -22,70 +22,202 @@
...
-    m_pGPUPerfAPI = NULL;
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, "Failed to get GPA function table. Invalid dynamic library?");
+    return false;
+  }
+
+  GPA_Logging_Type loggingType = GPA_LOGGING_ERROR;
+#if ENABLED(RDOC_DEVEL)
+  loggingType = GPA_LOGGING_ERROR_AND_MESSAGE;
+#endif
+  auto status = m_pGPUPerfAPI->GPA_RegisterLoggingCallback(loggingType, GPA_LoggingCallback);
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp:182]: `auto` -> `GPA_Status` again please.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 12]
```diff
@@ -22,70 +22,202 @@
...
+    message = "Info: ";
+    break;
+  case GPA_LOGGING_TRACE:
+    message = "Trace: ";
+    break;
+  default:
+    break;
+  }
+
+  message += "AMD: ";
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp]: Similarly I don't think there's a need for an extra prefix, this will already be noted as coming from `amd_counters.cpp`.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 13]
```diff
@@ -22,70 +22,202 @@
...
+inline bool AMD_SUCCEEDED(GPA_Status status)
+{
+  return status == GPA_STATUS_OK;
+}
+
+static void GPA_LoggingCallback(GPA_Logging_Type messageType, const char* pMessage)
+{
+  std::string message;
+
+  switch (messageType)
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp]: It ends up being redundant adding your own prefix here since it's then also prefixed with the message type by the pre-existing logging. If you wanted you could convert `GPA_LOGGING_ERROR` -> `RDCERR`, `GPA_LOGGING_MESSAGE` -> `RDCLOG` and `GPA_LOGGING_TRACE` -> `RDCDEBUG` but otherwise you should be able to print the message straight.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 14]
```diff
@@ -22,70 +22,202 @@
...
+
+  message.append(pMessage);
+
+  if (messageType == GPA_LOGGING_ERROR)
+  {
+    RDCWARN(message.c_str());
+  }
+  else
+  {
+    RDCLOG(message.c_str());
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp:74]: Careful not to pass a string like this that could in theory have printf formatters left in by accident or coincidence, instead do something like `RDCLOG("%s", message.c_str())` or as discussed above probably just pass `pMessage` directly.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 15]
```diff
@@ -22,70 +22,202 @@
...
+  }
+
+  GPA_Logging_Type loggingType = GPA_LOGGING_ERROR;
+#if ENABLED(RDOC_DEVEL)
+  loggingType = GPA_LOGGING_ERROR_AND_MESSAGE;
+#endif
+  auto status = m_pGPUPerfAPI->GPA_RegisterLoggingCallback(loggingType, GPA_LoggingCallback);
+  if (AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Failed to initialize logging", status).c_str());
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp]: It seems a little convoluted to go via `GPA_LoggingCallback` and as discussed there it should probably just forward this to `RDCERR`, so I think instead it'd be better to convert to something like:

```cpp
RDCERR("Failed to initialize logging. %s", m_pGPUPerfAPI->GPA_GetStatusAsStr(status));
```

If you wanted you could define a local macro to remove repetition a bit.

File: renderdoc/driver/ihv/amd/amd_counters.h [hunk 16]
```diff
@@ -25,20 +25,29 @@
...
 class AMDCounters
 {
 public:
+  enum ApiType
+  {
+    eApiType_Dx11 = 0,
+    eApiType_Dx12 = 1,
+    eApiType_Ogl = 2,
+    eApiType_Vk = 3
+  };
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.h:46]: It's not a hard rule but I tend to prefer `enum class` now that the code is compiled as C++11, so e.g. `enum class AMDCounterType { D3D11, D3D12, GL, Vulkan }` at global scope, or else you could re-use e.g. `enum class GraphicsAPI` which already has these values defined for you, since this doesn't have to match up to an internal enum.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 17]
```diff
@@ -101,34 +233,61 @@ AMDCounters::~AMDCounters()
...
+    CounterDescription desc = InternalGetCounterDescription(i);
+
+    desc.counter = MakeAMDCounter(i);
+    counters[i] = desc;
+
+    m_PublicToInternalCounter[desc.counter] = i;
   }
 
-  return counters;
+  return std::move(counters);
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp:277]: I'm not sure why the use of `std::move` here, since the function is still returning a temporary rather than a move reference. If you're worried about the copy cost I'd say instead define the function as `void AMDCounters::EnumerateCounters()` and write directly to `m_Counters` instead of a temporary, since with the addition of `m_PublicToInternalCounter` it's not re-usable somewhere else anyway.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 18]
```diff
@@ -227,140 +414,374 @@ CounterDescription AMDCounters::InternalGetCounterDescription(uint32_t internalI
...
+  const std::vector<GPUCounter> &counters
+)
+{
+  std::vector<CounterResult> ret;
+
+  bool isReady = false;
+
+  const uint32_t timeout = 10000; // ms
+
+  auto startTime = std::chrono::high_resolution_clock::now();
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp]: `PerformanceTimer` gives an abstracted timer, which should be used here instead of STL constructs.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 19]
```diff
@@ -227,140 +414,374 @@ CounterDescription AMDCounters::InternalGetCounterDescription(uint32_t internalI
...
+
+  do
+  {
+    isReady = IsSessionReady(sessionID);
+    if (!isReady)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(0));
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp]: Similarly, `Threading::Sleep` here instead of STL.

File: renderdoc/driver/ihv/amd/amd_counters.cpp [hunk 20]
```diff
@@ -227,140 +414,374 @@ CounterDescription AMDCounters::InternalGetCounterDescription(uint32_t internalI
...
+
+        ret.push_back(CounterResult(eventIDs[s], counters[c], value));
+      }
+      break;
+      default: RDCASSERT(0); break;
+      };
+    }
+  }
+
+  return std::move(ret);
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.cpp:564]: I think as discussed in `EnumerateCounters` all uses of `std::move` like this should be removed, the functions aren't returning rvalue references so the copy will still happen, leaving this confusing/misleading about what it does.

File: renderdoc/driver/ihv/amd/amd_counters.h [hunk 21]
```diff
@@ -1,7 +1,7 @@
 /******************************************************************************
  * The MIT License (MIT)
  *
- * Copyright (c) 2015-2018 Baldur Karlsson
+ * Copyright (c) 2015-2017 Baldur Karlsson
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.h]: Unless I missed something major, time isn't going backwards :wink: .

File: renderdoc/driver/vulkan/vk_debug.cpp [hunk 22]
```diff
@@ -1393,6 +1395,23 @@ void VulkanReplay::CreateResources()
       });
 
   CREATE_OBJECT(m_MeshFetchDescSet, m_General.DescriptorPool, m_MeshFetchDescSetLayout);
+
+  GPA_vkContextOpenInfo context{
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_debug.cpp]: Please don't use `{}` braces here when constructing an object, instead use normal `()` brackets.

File: renderdoc/driver/vulkan/vk_win32.cpp [hunk 23]
```diff
@@ -101,6 +101,17 @@ void WrappedVulkan::AddRequiredExtensions(bool instance, vector<string> &extensi
...
+    if (supportedExtensions.find(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME) == supportedExtensions.end())
+    {
+      RDCWARN("Unsupported required instance extension for AMD performance counters '%s'", VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+    }
+    else
+    {
+      if (std::find(extensionList.begin(), extensionList.end(),
+        VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME) == extensionList.end())
+        extensionList.push_back(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
+    }
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_win32.cpp]: Since this is needed on both platforms it should instead be done outside these files in `vk_device_funcs.cpp` in `WrappedVulkan::Serialise_vkCreateDevice`, the same place where `EXT_debug_marker` and `AMD_shader_info` are enabled.

Also as a follow up, do we need to store a flag somewhere as to whether this extension was enabled? So we can check it later to gate any use that might depend on it. I don't see anything directly so I assume it's inside the GPA libraries themselves?

File: renderdoc/driver/ihv/amd/amd_counters.h [hunk 24]
```diff
@@ -25,20 +25,29 @@
 #pragma once
 
 #include <vector>
+#include <map>
 #include "api/replay/renderdoc_replay.h"
+#include "api/replay/data_types.h"
```
Comment (baldurk) [inline renderdoc/driver/ihv/amd/amd_counters.h]: There shouldn't be a need to directly include `data_types.h` here - it's included under `renderdoc_replay.h`. Generally external code should only ever include the root `renderdoc_replay.h` and not any of the children since there are some sensitive ordering issues and declarations may move around.

File: renderdoc/driver/vulkan/vk_counters.cpp [hunk 25]
```diff
@@ -22,9 +22,13 @@
  * THE SOFTWARE.
  ******************************************************************************/
 
+#include <algorithm>
+#include <iterator>
 #include "vk_core.h"
 #include "vk_replay.h"
 #include "vk_resources.h"
+#include "driver/ihv/amd/amd_counters.h"
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_counters.cpp]: You might fall afoul of clang-format telling you to move these headers above `vk_core.h` - if you need to maintain this order just add a space so they're two separate groups of includes.

**Review by baldurk — CHANGES_REQUESTED** [review]
Thanks for this change! It looks great and will be a useful feature to have in for v1.0.

I've made comments to review the code. I don't think there's anything requiring large restructuring but mostly tidy up, clarification, and local re-organisation.

Usually I don't mind how people organise commits on their branch but in this case you have one commit adding something then the next is pretty much reverting it, so in that case I'd ask that you squash the changes from the second into the first. That way the history is tidier. You'll also need to apply clang-format and squash that in as well.

File: qrenderdoc/qrenderdoc_local.vcxproj (hunk 1 above)
Comment (herb-amd) [reply qrenderdoc/qrenderdoc_local.vcxproj:1649]: We had to do this starting with an integration we picked up in mid-December.  Reverting as current code works without it.

File: renderdoc/data/hlsl/debugdisplay.hlsl (hunk 2 above)
Comment (herb-amd) [reply renderdoc/data/hlsl/debugdisplay.hlsl:203]: This fixes a shader error visualizing certain Direct3D12 games.

File: renderdoc/driver/d3d11/d3d11_replay.cpp (hunk 3 above)
Comment (herb-amd) [reply renderdoc/driver/d3d11/d3d11_replay.cpp]: Fixed

File: renderdoc/driver/d3d12/d3d12_counters.cpp (hunk 4 above)
Comment (herb-amd) [reply renderdoc/driver/d3d12/d3d12_counters.cpp]: Fixed

File: renderdoc/driver/d3d12/d3d12_counters.cpp (hunk 5 above)
Comment (herb-amd) [reply renderdoc/driver/d3d12/d3d12_counters.cpp]: The issue is that we can't begin sampling on a command list until we know that it contains at least one sample; and, we don't know that until we get the draw.

File: renderdoc/driver/ihv/amd/amd_counters.h (hunk 21 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.h]: Fixed

File: renderdoc/driver/vulkan/vk_debug.cpp (hunk 22 above)
Comment (herb-amd) [reply renderdoc/driver/vulkan/vk_debug.cpp]: This is a struct without any ctor.

File: renderdoc/driver/vulkan/vk_win32.cpp (hunk 23 above)
Comment (herb-amd) [reply renderdoc/driver/vulkan/vk_win32.cpp]: Fixed.  It does not need to be saved.  If not found, it informs the user why they won't see any performance counters.

File: renderdoc/driver/ihv/amd/amd_counters.h (hunk 24 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.h]: Fixed

File: renderdoc/driver/vulkan/vk_counters.cpp (hunk 25 above)
Comment (herb-amd) [reply renderdoc/driver/vulkan/vk_counters.cpp]: Fixed

File: renderdoc/data/hlsl/debugdisplay.hlsl (hunk 2 above)
Comment (baldurk) [reply renderdoc/data/hlsl/debugdisplay.hlsl:203]: Can you be more specific? I don't see how this declaration would interact with any of the capture shaders, or what you mean by visualizing exactly.

File: renderdoc/driver/d3d12/d3d12_counters.cpp (hunk 5 above)
Comment (baldurk) [reply renderdoc/driver/d3d12/d3d12_counters.cpp]: Fair enough! I just wondered if that would help, but this makes sense then.

File: renderdoc/driver/vulkan/vk_debug.cpp (hunk 22 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_debug.cpp]: OK, then it should be written as `GPA_vkContextOpenInfo context = { ... }` for C style brace initialization.

File: renderdoc/driver/vulkan/vk_win32.cpp (hunk 23 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_win32.cpp]: My concern is that most users will not look at a warning printed to the log file, or won't notice it in amongst other log prints. I think we should at least save this as a bool and check it before enumerating AMD counters (i.e. treat the situation where the extension is missing as the same as when the GPA interface fails to initialise). That way users might wonder why the AMD counters aren't available but they at least won't be shown counters that then don't work.

Comment (herb-amd) [reply renderdoc/driver/vulkan/vk_win32.cpp]: To clarify, the counters won't be populated.  This is the user's log file info as to why they weren't populated.

Comment (baldurk) [reply renderdoc/driver/vulkan/vk_win32.cpp]: Oh OK, sorry I misunderstood, that's fine then!

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 20 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp:564]: Previously fixed

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 18 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp]: Fixed

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 17 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp:277]: Previously fixed

File: renderdoc/driver/ihv/amd/amd_counters.h (hunk 16 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.h:46]: Fixed

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 14 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp:74]: Fixed

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 12 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp]: Fixed

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 11 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp:182]: Previously fixed

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 10 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp:127]: Fixed

File: renderdoc/driver/ihv/amd/amd_counters.cpp (hunk 8 above)
Comment (herb-amd) [reply renderdoc/driver/ihv/amd/amd_counters.cpp:84]: Fixed

File: renderdoc/driver/d3d12/d3d12_counters.cpp (hunk 6 above)
Comment (herb-amd) [reply renderdoc/driver/d3d12/d3d12_counters.cpp:451]: Fixed
